linkerd/http/detect/src/metrics.rs

```diff
@@ -0,0 +1,195 @@
+    results: Family<DetectLabels<L>, Counter>,
+}
+
+#[derive(Clone, Debug)]
+pub struct DetectMetrics {
+    duration: Histogram,
+    not_http: Counter,
+    http1: Counter,
+    h2: Counter,
+    read_timeout: Counter,
+    error: Counter,
+}
+
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+struct DetectLabels<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSet + Debug + Send + Sync + 'static,
+{
+    result: DetectResult,
+    labels: L,
+}
+
+#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
+enum DetectResult {
+    NotHttp,
+    Http1,
+    H2,
+    ReadTimeout,
+    Error,
+}
+
+#[derive(Clone, Debug, Default)]
+struct MkDurations;
+
+// === impl DetectMetricsFamilies ===
+
+impl<L> Default for DetectMetricsFamilies<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSet + Debug + Send + Sync + 'static,
+{
+    fn default() -> Self {
+        Self {
+            duration: Family::new_with_constructor(MkDurations),
+            results: Family::default(),
+        }
+    }
+}
+
+impl<L> DetectMetricsFamilies<L>
+where
+    L: Clone + Hash + Eq + EncodeLabelSet + Debug + Send + Sync + 'static,
+{
+    pub fn register(reg: &mut Registry) -> Self {
+        let duration = Family::new_with_constructor(MkDurations);
+        reg.register_with_unit(
+            "duration",
+            "Time taken for protocol detection",
+            Unit::Seconds,
+            duration.clone(),
+        );
+
+        let results = Family::default();
+        reg.register("results", "Protocol detection results", results.clone());
+
+        Self { duration, results }
+    }
+
+    pub fn metrics(&self, labels: L) -> DetectMetrics {
+        let duration = (*self.duration.get_or_create(&labels)).clone();
+
+        let not_http = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::NotHttp,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let http1 = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::Http1,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let h2 = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::H2,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let read_timeout = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::ReadTimeout,
+            labels: labels.clone(),
+        }))
+        .clone();
+        let error = (*self.results.get_or_create(&DetectLabels {
+            result: DetectResult::Error,
+            labels,
+        }))
+        .clone();
+
+        DetectMetrics {
+            duration,
+            not_http,
+            http1,
+            h2,
+            read_timeout,
+            error,
+        }
+    }
+}
+
+// === impl DetectMetrics ===
+
+impl Default for DetectMetrics {
+    fn default() -> Self {
+        Self {
+            duration: Histogram::new([0.001, 0.1].into_iter()),
```

Comment:
nit: would it make sense to use `MkDurations::new_metric()` here? these share a `[0.001, 0.1]` for the buckets.